{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import sys, os\n",
    "sys.path.append(\"../scripts\")\n",
    "\n",
    "import logging\n",
    "logging.basicConfig()\n",
    "logger = logging.getLogger(__name__)\n",
    "logger.setLevel('INFO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jsi17/anaconda3/envs/tf2-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3170: DtypeWarning: Columns (5) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "INFO:__main__:genotypes shape is (1814, 10346), phenotypes shape is (1814, 6)\n"
     ]
    }
   ],
   "source": [
    "# load the data\n",
    "genotypes = pd.read_csv(\"X.csv\", index_col=0)\n",
    "phenotypes = pd.read_csv(\"phenotypes.csv\", index_col=0)\n",
    "snp_locations = pd.read_csv(\"mice_map2.csv\")\n",
    "marker_locations = pd.read_csv(\"MGI_MRK_Coord.rpt.csv\")\n",
    "marker_locations.Chromosome = marker_locations.Chromosome.astype(str)\n",
    "\n",
    "logger.info(\"genotypes shape is {}, phenotypes shape is {}\".format(genotypes.shape, phenotypes.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Marker Type</th>\n",
       "      <th>Feature Type</th>\n",
       "      <th>Marker Symbol</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start Coordinate</th>\n",
       "      <th>End Coordinate</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Genome Build</th>\n",
       "      <th>Provider Collection</th>\n",
       "      <th>Provider Display</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:87860</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>Abl2</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene 2...</td>\n",
       "      <td>1</td>\n",
       "      <td>156558786</td>\n",
       "      <td>156649568</td>\n",
       "      <td>+</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>Ensembl Gene Model</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:87866</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>Acadl</td>\n",
       "      <td>acyl-Coenzyme A dehydrogenase, long-chain</td>\n",
       "      <td>1</td>\n",
       "      <td>66830839</td>\n",
       "      <td>66863277</td>\n",
       "      <td>-</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>Ensembl Gene Model</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:87893</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>Chrnd</td>\n",
       "      <td>cholinergic receptor, nicotinic, delta polypep...</td>\n",
       "      <td>1</td>\n",
       "      <td>87190607</td>\n",
       "      <td>87200070</td>\n",
       "      <td>+</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>Ensembl Gene Model</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:87895</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>Chrng</td>\n",
       "      <td>cholinergic receptor, nicotinic, gamma polypep...</td>\n",
       "      <td>1</td>\n",
       "      <td>87205811</td>\n",
       "      <td>87211835</td>\n",
       "      <td>+</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>NCBI Gene Model</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:87948</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>Adss</td>\n",
       "      <td>adenylosuccinate synthetase, non muscle</td>\n",
       "      <td>1</td>\n",
       "      <td>177763176</td>\n",
       "      <td>177796695</td>\n",
       "      <td>-</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>NCBI Gene Model</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265365</th>\n",
       "      <td>MGI:6092513</td>\n",
       "      <td>Other Genome Feature</td>\n",
       "      <td>TSS region</td>\n",
       "      <td>Tssr164746</td>\n",
       "      <td>transcription start site region 164746</td>\n",
       "      <td>Y</td>\n",
       "      <td>4225995</td>\n",
       "      <td>4226013</td>\n",
       "      <td>-</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>RIKEN</td>\n",
       "      <td>RIKEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265366</th>\n",
       "      <td>MGI:6092514</td>\n",
       "      <td>Other Genome Feature</td>\n",
       "      <td>TSS region</td>\n",
       "      <td>Tssr164747</td>\n",
       "      <td>transcription start site region 164747</td>\n",
       "      <td>Y</td>\n",
       "      <td>4226100</td>\n",
       "      <td>4226106</td>\n",
       "      <td>-</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>RIKEN</td>\n",
       "      <td>RIKEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265367</th>\n",
       "      <td>MGI:6092515</td>\n",
       "      <td>Other Genome Feature</td>\n",
       "      <td>TSS region</td>\n",
       "      <td>Tssr164748</td>\n",
       "      <td>transcription start site region 164748</td>\n",
       "      <td>Y</td>\n",
       "      <td>4227178</td>\n",
       "      <td>4227185</td>\n",
       "      <td>-</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>RIKEN</td>\n",
       "      <td>RIKEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265368</th>\n",
       "      <td>MGI:6096127</td>\n",
       "      <td>Gene</td>\n",
       "      <td>miRNA gene</td>\n",
       "      <td>Gm47281</td>\n",
       "      <td>predicted gene, 47281</td>\n",
       "      <td>Y</td>\n",
       "      <td>991657</td>\n",
       "      <td>991716</td>\n",
       "      <td>-</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>Ensembl Gene Model</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265369</th>\n",
       "      <td>MGI:6096131</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>Gm47283</td>\n",
       "      <td>predicted gene, 47283</td>\n",
       "      <td>Y</td>\n",
       "      <td>90784738</td>\n",
       "      <td>90816464</td>\n",
       "      <td>+</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>Ensembl Gene Model</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>265370 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       MGI Marker Accession ID           Marker Type         Feature Type  \\\n",
       "0                    MGI:87860                  Gene  protein coding gene   \n",
       "1                    MGI:87866                  Gene  protein coding gene   \n",
       "2                    MGI:87893                  Gene  protein coding gene   \n",
       "3                    MGI:87895                  Gene  protein coding gene   \n",
       "4                    MGI:87948                  Gene  protein coding gene   \n",
       "...                        ...                   ...                  ...   \n",
       "265365             MGI:6092513  Other Genome Feature           TSS region   \n",
       "265366             MGI:6092514  Other Genome Feature           TSS region   \n",
       "265367             MGI:6092515  Other Genome Feature           TSS region   \n",
       "265368             MGI:6096127                  Gene           miRNA gene   \n",
       "265369             MGI:6096131                  Gene  protein coding gene   \n",
       "\n",
       "       Marker Symbol                                        Marker Name  \\\n",
       "0               Abl2  v-abl Abelson murine leukemia viral oncogene 2...   \n",
       "1              Acadl          acyl-Coenzyme A dehydrogenase, long-chain   \n",
       "2              Chrnd  cholinergic receptor, nicotinic, delta polypep...   \n",
       "3              Chrng  cholinergic receptor, nicotinic, gamma polypep...   \n",
       "4               Adss            adenylosuccinate synthetase, non muscle   \n",
       "...              ...                                                ...   \n",
       "265365    Tssr164746             transcription start site region 164746   \n",
       "265366    Tssr164747             transcription start site region 164747   \n",
       "265367    Tssr164748             transcription start site region 164748   \n",
       "265368       Gm47281                              predicted gene, 47281   \n",
       "265369       Gm47283                              predicted gene, 47283   \n",
       "\n",
       "       Chromosome  Start Coordinate  End Coordinate Strand Genome Build  \\\n",
       "0               1         156558786       156649568      +       GRCm38   \n",
       "1               1          66830839        66863277      -       GRCm38   \n",
       "2               1          87190607        87200070      +       GRCm38   \n",
       "3               1          87205811        87211835      +       GRCm38   \n",
       "4               1         177763176       177796695      -       GRCm38   \n",
       "...           ...               ...             ...    ...          ...   \n",
       "265365          Y           4225995         4226013      -       GRCm38   \n",
       "265366          Y           4226100         4226106      -       GRCm38   \n",
       "265367          Y           4227178         4227185      -       GRCm38   \n",
       "265368          Y            991657          991716      -       GRCm38   \n",
       "265369          Y          90784738        90816464      +       GRCm38   \n",
       "\n",
       "       Provider Collection Provider Display  \n",
       "0       Ensembl Gene Model          Ensembl  \n",
       "1       Ensembl Gene Model          Ensembl  \n",
       "2       Ensembl Gene Model          Ensembl  \n",
       "3          NCBI Gene Model             NCBI  \n",
       "4          NCBI Gene Model             NCBI  \n",
       "...                    ...              ...  \n",
       "265365               RIKEN            RIKEN  \n",
       "265366               RIKEN            RIKEN  \n",
       "265367               RIKEN            RIKEN  \n",
       "265368  Ensembl Gene Model          Ensembl  \n",
       "265369  Ensembl Gene Model          Ensembl  \n",
       "\n",
       "[265370 rows x 12 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marker_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:__main__:8854 of 10346 genotypes have locations\n"
     ]
    }
   ],
   "source": [
    "# some formatting of the genotypes\n",
    "new_snp_names = [x[0] for x in genotypes.columns.str.split(\"_\").str[:1]]\n",
    "genotypes.columns = new_snp_names\n",
    "\n",
    "snps_with_locs = np.isin(genotypes.columns, snp_locations.marker)\n",
    "logger.info(f\"{snps_with_locs.sum()} of {genotypes.shape[1]} genotypes have locations\")\n",
    "\n",
    "genotypes = genotypes.loc[:,snps_with_locs]\n",
    "snp_locations = snp_locations.loc[ snp_locations.marker.isin(genotypes.columns) ].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# map the SNPs to the markers (Genes) by iterating over the SNPs\n",
    "# and seeing if they fall inside the marker boundaries\n",
    "# this is very slow, takes around 10 minutes on my machine\n",
    "# SNPs can be on multiple markers as they overlap\n",
    "snp2marker = {}\n",
    "\n",
    "for _, snp_row in snp_locations.iterrows():\n",
    "    markers_on_chromosome = marker_locations.loc[ marker_locations.Chromosome==snp_row.chromosome ]\n",
    "    row_indexer = np.logical_and(\n",
    "        markers_on_chromosome[\"Start Coordinate\"] < snp_row.bp,\n",
    "        markers_on_chromosome[\"End Coordinate\"] > snp_row.bp\n",
    "    )\n",
    "    markers_matching_snp = markers_on_chromosome.loc[row_indexer,:]\n",
    "    if markers_matching_snp.shape[0] > 1:\n",
    "        snp2marker[snp_row.marker] = markers_matching_snp[\"Marker Symbol\"].to_numpy().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>Marker</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3707673</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs3707673</td>\n",
       "      <td>Gm37180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs3707673</td>\n",
       "      <td>Gm39585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs3707673</td>\n",
       "      <td>Ebas2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3707673</td>\n",
       "      <td>Twq5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121993</th>\n",
       "      <td>gnfX.148.995</td>\n",
       "      <td>Obsty5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121994</th>\n",
       "      <td>rs13484113</td>\n",
       "      <td>Li</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121995</th>\n",
       "      <td>rs13484113</td>\n",
       "      <td>Bw3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121996</th>\n",
       "      <td>rs13484113</td>\n",
       "      <td>Spha3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121997</th>\n",
       "      <td>rs13484113</td>\n",
       "      <td>Obsty5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121998 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 snp   Marker\n",
       "0          rs3707673     Xkr4\n",
       "1          rs3707673  Gm37180\n",
       "2          rs3707673  Gm39585\n",
       "3          rs3707673    Ebas2\n",
       "4          rs3707673     Twq5\n",
       "...              ...      ...\n",
       "121993  gnfX.148.995   Obsty5\n",
       "121994    rs13484113       Li\n",
       "121995    rs13484113      Bw3\n",
       "121996    rs13484113    Spha3\n",
       "121997    rs13484113   Obsty5\n",
       "\n",
       "[121998 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_dict_map = pd.concat(\n",
    "    {k: pd.Series(v) for k, v in snp2marker.items()}\n",
    ").reset_index().rename(\n",
    "    columns={'level_0' : 'snp', 0 : 'Marker'}\n",
    ").drop(\n",
    "    columns=\"level_1\"\n",
    ")\n",
    "snp_dict_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-11 15:29:49 INFO     Keeping 5,503 SNPs and 764 genes\n"
     ]
    }
   ],
   "source": [
    "# some post processing - we are only interested in SNPs on a single\n",
    "# gene\n",
    "snps_per_marker = snp_dict_map.groupby(\"Marker\", as_index=False).size()\n",
    "snps_per_marker.sort_values(\"size\", ascending=False)\n",
    "\n",
    "# only keep SNPs mapped to genes\n",
    "genes_only_map = snp_dict_map.loc[ snp_dict_map.Marker.isin(\n",
    "    marker_locations[\"Marker Symbol\"][ marker_locations[\"Marker Type\"]==\"Gene\" ]\n",
    ")].reset_index(drop=True)\n",
    "\n",
    "# only keep genes with more than one SNP\n",
    "genes_only_map = genes_only_map.loc[ genes_only_map.Marker.isin(\n",
    "    snps_per_marker[\"Marker\"][ snps_per_marker[\"size\"]>1 ]\n",
    ")]\n",
    "\n",
    "included_snps = genes_only_map.snp.unique()\n",
    "included_genes = genes_only_map.Marker.unique()\n",
    "\n",
    "logger.info(f\"Keeping {included_snps.shape[0]:,} SNPs and {included_genes.shape[0]} genes\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tf2-gpu",
   "language": "python",
   "name": "tf2-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
